### ISSUE TITLE
Middleware not working when accessing directly to a auth route

### ISSUE THREAD

**Issue by DannyAndres** [razorcx-courses/nuxt3-firebase-auth#1]
If I access /home from url as the first page to load it will not find auth variable with 

`const auth = getAuth();`
`nuxtApp.vueApp.provide('auth', auth);`
`nuxtApp.provide('auth', auth);`

because it cannot find it, it will not get passed to the middleware behaving as if there is no user at all.

When the user is already loaded everything work as expected, I suspect it is getAuth() which does not load as quick as it should causing all of this.

**Comment by minstn** [comment]
i guess you mean `/secret` page not `/home`? in fact, the "logged in user" is only known after the `onAuthStateChanged` is called.

**Comment by DannyAndres** [comment]
yeah, /secret and /home in this case is to represent a route which should be protected and only accesible with a logged user. I think the problem comes with `onAuthStateChanged` takes some time to load at the beginning

**Comment by silvanm** [comment]
> I think the problem comes with `onAuthStateChanged` takes some time to load at the beginning

I agree that this is the reason. 

Is there a way to wait for `onAuthStateChanged` to be called on before the Middlewares are being called? 

Or should we store the current user in localStorage so that we're independent of `onAuthStateChanged`?

**Comment by smoakey** [comment]
@silvanm did you ever find an answer to your problem above?

**Comment by Eronne** [comment]
Same issue here! Middleware are triggered before the user is authenticated...

**Comment by superdiazzz** [comment]
I have strange issue in my middleware, when i console log like below:
```
export default defineNuxtPlugin(() => {
    addRouteMiddleware('global-test', () => {

    }, { global: true })

    addRouteMiddleware('auth', async() => {

        const { $auth } = useNuxtApp()


        console.log("call auth middle 1", $auth);

        console.log('call auth middle 2', $auth.currentUser);  
    })
})
```

first console shows there is a value, but the second is null??

![Screen Shot 2024-01-21 at 16 49 42](https://github.com/razorcx-courses/nuxt3-firebase-auth/assets/10770867/b580232b-271b-46a4-afc9-859a16e6b14c)

**Comment by minstn** [comment]
I assume the $auth.currentUser is not yet initialized, as Firebase Auth is
a long process which happens in the browser, not on the server.

**Comment by EntertainmentPortal** [comment]
Found the solution:

- Remove the plugins/auth.js
- Create a folder called 'middleware'
- Create a file inside middleware called 'auth.js'
- Add this code to middleware/auth.js:
`export default defineNuxtRouteMiddleware((to, from) => {
  const user = useFirebaseUser();
  const allowedRoutes = ["/", "/login"];
  if (user.value) {
    if (allowedRoutes.includes(to.path)) {
      return navigateTo("/secret");
    }
  } else {
    if (!allowedRoutes.includes(to.path)) {
      return navigateTo("/");
    }
  }
});
`
- Update the allowed routes accordingly to your page structure
- Remove the 'middleware: ['auth']' from definePageMeta on every page.
- That's it.

What it does: it defines the middleware globally, guarding all routes. If you want any exceptions, you can add them to the allowed list. It checks for the useFirebaseUser, which contains a watcher too
